src/model64.c

```diff
@@ -740,38 +688,185 @@ static void quat_lerp(float *out, float *in1, float *in2, float time)
+                    decoded_values[j][3] = 0.0f;
+                    break;
+                    
+                case ANIM_COMPONENT_ROT:
+                    decode_quaternion(frame->data, &decoded_values[j][0]);
+                    break;
+                    
+                case ANIM_COMPONENT_SCALE:
+                    decode_normalized_u16(frame->data, &decoded_values[j][0], 3, curr_anim->scale_min, curr_anim->scale_max);
+                    decoded_values[j][3] = 0.0f;
+                    break;
+                  
+                default:
+                    decoded_values[j][0] = decoded_values[j][1] = decoded_values[j][2] = decoded_values[j][3] = 0.0f;
+                    break;
+            }
+        }
+        float time = anim_state->frames[(i*4)+((1+anim_state->buf_idx[i]) % 4)].time;
+        float time_next = anim_state->frames[(i*4)+((2+anim_state->buf_idx[i]) % 4)].time;
+        float weight;
+        if(anim_state->time > time_next) {
+            weight = 1.0f;
+        } else {
+            if(anim_state->time < time) {
+                weight = 0.0f;
+            } else {
+                weight = (anim_state->time-time)/(time_next-time);
+            }
+            
+        }
+        float *out;
+        size_t out_count;
         switch(component) {
             case ANIM_COMPONENT_POS:
-                vec_lerp(model->transforms[node_index].transform.pos, (float *)curr_buf, (float *)next_buf, 3, anim_buf_info.time);
-                data_size = 3*sizeof(float); 
+                out = model->transforms[node].transform.pos;
+                out_count = 3;
                 break;
                 
             case ANIM_COMPONENT_ROT:
-                quat_lerp(model->transforms[node_index].transform.rot, (float *)curr_buf, (float *)next_buf, anim_buf_info.time);
-                data_size = 4*sizeof(float); 
+                out = model->transforms[node].transform.rot;
+                out_count = 4;
                 break;
                 
             case ANIM_COMPONENT_SCALE:
-                vec_lerp(model->transforms[node_index].transform.scale, (float *)curr_buf, (float *)next_buf, 3, anim_buf_info.time);
-                data_size = 3*sizeof(float); 
+                out = model->transforms[node].transform.scale;
+                out_count = 3;
+                break;
+              
+            default:
+                out = NULL;
                 break;
         }
-        calc_node_local_matrix(model, node_index);
-        curr_buf += data_size;
-        next_buf += data_size;
+        if(out == NULL) {
+            continue;
+        }
+        if(component == ANIM_COMPONENT_ROT) {
+            quat_lerp(out, &decoded_values[1][0], &decoded_values[2][0], weight);
+        } else {
+            catmull_calc_vec(&decoded_values[0][0], &decoded_values[1][0], &decoded_values[2][0], &decoded_values[3][0], out, weight, out_count);
+        }
+        calc_node_local_matrix(model, node);
```

Comment:
Shouldn't this be done just once per node? It looks like it's done once per done per component.

Reply:
It could be done that way if one sorts the tracks.